Install latest release for an Xcode plugin

neonichu opened the pull request:
Using the `github` option instead of `url`, fastlane can now determine the URL for the latest version of a plugin automatically.

neonichu committed:
Install latest release for an Xcode plugin

Using the `github` option instead of `url`, fastlane can now determine
the URL for the latest version of a plugin automatically.

KrauseFx commented on fastlane/docs/Actions.md:
````diff
@@ -2438,6 +2438,12 @@ Install an Xcode plugin for the current user
 install_xcode_plugin(url: 'https://github.com/contentful/ContentfulXcodePlugin/releases/download/0.5/ContentfulPlugin.xcplugin.zip')
 ```
 
+You can also let fastlane pick the latest version of a plugin automatically, if it is hosted on GitHub
````
Should be `fastlane` with ` around the word :+1:

KrauseFx commented on fastlane/docs/Actions.md:
```diff
@@ -2438,6 +2438,12 @@ Install an Xcode plugin for the current user
 install_xcode_plugin(url: 'https://github.com/contentful/ContentfulXcodePlugin/releases/download/0.5/ContentfulPlugin.xcplugin.zip')
```
Let's use a different example for the `url` then, so that it's clear to not use this for GitHub links

neonichu replied:
Makes sense, do you think it's OK to use a fake URL? Don't actually know of any plugins not hosted on :octocat:

KrauseFx replied:
Yeah, let's use a really obvious fake URL, like `https://randomurl.com/plugin.zip`

KrauseFx commented on fastlane/lib/fastlane/actions/install_xcode_plugin.rb:
```diff
@@ -4,6 +4,12 @@ class InstallXcodePluginAction < Action
       def self.run(params)
         require 'fileutils'
 
+        if params.fetch(:github, ask: false)
```
Instead just define this option as `optional: true`, then you can just use `params[:github]`

KrauseFx commented on fastlane/lib/fastlane/actions/install_xcode_plugin.rb:
```diff
@@ -4,6 +4,12 @@ class InstallXcodePluginAction < Action
+        if params.fetch(:github, ask: false)
+          github_api_url = params[:github].sub('https://github.com', 'https://api.github.com/repos')
+          release = self.fetch_json(github_api_url + '/releases/latest')
+          params[:url] = release['assets'][0]['browser_download_url']
```
This will crash with a cryptic error when something's wrong, or there are no releases, right?

neonichu replied:
YES! ✨

neonichu replied:
Showing a generic error when GH errors and a specific one for missing latest release now.

KrauseFx commented on fastlane/lib/fastlane/actions/install_xcode_plugin.rb:
```diff
@@ -30,6 +42,13 @@ def self.available_options
+                                       env_name: "FL_XCODE_PLUGIN_GITHUB",
+                                       description: "GitHub repository URL for Xcode plugin",
+                                       verify_block: proc do |value|
+                                         UI.user_error!("No GitHub URL for InstallXcodePluginAction given, pass using `github: 'url'`") if value.to_s.length == 0
```
Remove this check and define this ConfigItem to be `optional: true` instead

neonichu committed:
Better error handling and minor fixes

- Handles non-existent releases and other GH errors
- Make `:github` option optional
- Fix formatting and use fake URL for `url` in `Actions.md`

neonichu commented:
🔑

KrauseFx commented:
This looks great, however no tests. Is that something you tested locally to make sure it works? Do you mind adding 1 or 2 tests for it?

neonichu commented:
I've tested locally, of course 😉

I was thinking about adding some tests, but wasn't sure about the approach to take — this is shelling out a lot — maybe testing the interaction with he GH API could have value, though?

KrauseFx commented:
Hey @neonichu, sorry danger is broken, could you rebase from master and push the changes?

neonichu commented:
rebased ✅

KrauseFx commented on fastlane/docs/Actions.md:
````diff
@@ -2435,7 +2435,13 @@ fastlane_version "1.50.0"
 
 ```ruby
-install_xcode_plugin(url: 'https://github.com/contentful/ContentfulXcodePlugin/releases/download/0.5/ContentfulPlugin.xcplugin.zip')
+install_xcode_plugin(url: 'https://example.com/clubmate/plugin.zip')
````
club mate, best!

KrauseFx commented on fastlane/docs/Actions.md:
````diff
@@ -2435,7 +2435,13 @@ fastlane_version "1.50.0"
+You can also let `fastlane` pick the latest version of a plugin automatically, if it is hosted on GitHub
+
+```ruby
+install_xcode_plugin(github: 'https://github.com/contentful/ContentfulXcodePlugin')
````
We should actually use `install_xcode_plugin(github: 'https://github.com/fastlane/FixCode')` as example ^^

KrauseFx commented:
This is great, thanks @neonichu, I also PRd https://github.com/neonichu/FixCode/pull/18

KrauseFx commented:
Thanks 🚀
